text/0615-autotracking-memoization.md

```diff
@@ -0,0 +1,385 @@
+nested, and whenever you use a cache inside of another cache, the outer cache
+will dirty if the inner cache dirties.
+
+```ts
+let inner = createCache(() => { /* ... */ })
+
+let outer = createCache(() => {
+  /* ... */
+
+  inner();
+});
+```
+
+This can be used to break up different parts of a execution so that only the
+pieces that changed are rerun.
+
+The `isCache` function can be used to determine if a value is a cache, in cases
+where it's not possible to know if the value is a cache or not.
+
+```js
+let cache = createCache(() => {});
+let notACache = () => {};
+
+isCache(cache); // true
+isCache(notACache); // false
+```
+
+### Constant Caches
+
+Caches will only recompute if any of the tracked inputs that were consumed
+previously change. If there _were_ no consumed tracked inputs, then they will
+never recompute.
+
+```ts
+let count = 0;
+
+let counter = createCache(() => {
+  return ++count;
+});
+
+getValue(counter); // 1
+getValue(counter); // 1
+getValue(counter); // 1
+
+// ...
+```
+
+When this happens, it often means that optimizations can be made in the code
+surrounding the computation. For instance, in the Glimmer VM, we don't emit
+updating bytecodes if we detect that a memoized function is constant, because
+it means that this piece of DOM will never update.
+
+In order to check if a memoized function is constant or not, users can use the
+`isConst` function:
+
+```ts
+import { createCache, getValue, isConst } from '@glimmer/tracking/primitives/cache';
+
+class State {
+  @tracked value;
+}
+
+let state = new State();
+let count = 0;
+
+let counter = createCache(() => {
+  // consume the state
+  state.value;
+
+  return ++count;
+});
+
+
+let constCounter = createCache(() => {
+  return ++count;
+});
+
+getValue(counter);
+getValue(constCounter);
+
+isConst(counter); // false
+isConst(constCounter); // true
+```
+
+It is not possible to know whether or not a cache is constant before its
+first usage, so `isConst` will throw an error if the cache has never been
+accessed before.
+
+```ts
+
+let constCounter = createCache(() => {
+  return count++;
+});
+
+isConst(constCounter); // throws an error, `constCounter` has not been used
+```
+
+This helps users avoid missing optimization opportunities by mistake, since most
+optimizations happen on the first run only. If a user calls `isConst` on the
+function prior to the first run, they may assume that the function is
+non-constant on accident.
+
+## How we teach this
+
+This topic is one that is meant for advanced users and library authors. It
+should be covered in detail in the Autotracking In-Depth guide in the Ember
+guides.
+
+This guide should cover how memoization works, and various techniques for using
+memoization. It should cover a variety of ways to use memoization to accomplish
+common tasks of other reactivity systems. Pull-based reactivity is unfamiliar to
+many programmers, so we should try to familiarize them with as many common
+examples as possible.
```

Comment:
Agreed.  Even as we migrate applications over, I'm sometimes not applying `@tracked` in a condoned way